File: problems/chapter15/miiitomi/003.cpp
// https://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=1350

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct UnionFind {
    vector<int> par, siz;

    UnionFind(int n) : par(n, -1), siz(n, 1) {}

    int root(int x) {
        if (par[x] == -1) return x;
        else return par[x] = root(par[x]);
    }

    bool issame(int x, int y) {
        return root(x) == root(y);
    }

    bool unite(int x, int y) {
        x = root(x);
        y = root(y);

        if (x == y) return false;

        if (siz[x] < siz[y]) swap(x, y);

        par[y] = x;
        siz[x] += siz[y];
        return true;
    }

    int size(int x) {
        return siz[root(x)];
    }
};

int main() {
    int N, M;
    cin >> N >> M;
    vector<pair<int, pair<int, int>>> V(M);
    for (int i = 0; i < M; i++) {
        int u, v, c;
        cin >> u >> v >> c;
        u--;
        v--;
        V[i].first = c;
        V[i].second = make_pair(u, v);
    }
    sort(V.begin(), V.end());

    UnionFind uf(N);

    long long minimum_cost = 0LL;
    vector<pair<int, pair<int, int>>> W;
    for (int i = 0; i < M; i++) {
        if (uf.issame(V[i].second.first, V[i].second.second)) continue;
    
        minimum_cost += V[i].first;
        W.push_back(V[i]);
        uf.unite(V[i].second.first, V[i].second.second);
    }

    long long sum_cost = 0LL;
    int counter = 0;
    for (auto e : W) {
        UnionFind uf2(N);
        long long cost = 0LL;
        for (int i = 0; i < M; i++) {
            if (V[i] == e || uf2.issame(V[i].second.first, V[i].second.second)) continue;
            cost += V[i].first;
            uf2.unite(V[i].second.first, V[i].second.second);
        }
        if (uf2.size(0) != N || cost > minimum_cost) {
            sum_cost += e.first;
            counter++;
        }
    }

    cout << counter << " " << sum_cost << endl;
}

File: problems/chapter15/mshibatatt/003.py
# https://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=1350
class UnionFind():
    def __init__(self, n) -> None:
        self.par = [-1] * n
        self.siz = [1] * n
    
    # return root of x
    def root(self, x):
        if self.par[x] == -1:
            return x
        else: 
            self.par[x] = self.root(self.par[x])
            return self.par[x]
    
    # return if x and y are in the same group
    def issame(self, x, y):
        return self.root(x) == self.root(y)
    
    # combine group with x and one with y
    def unite(self, x, y):
        x = self.root(x)
        y = self.root(y)

        if x == y:
            return False
        if self.siz[x] < self.siz[y]:
            x, y = y, x
        
        # make y child of x
        self.par[y] = x
        self.siz[x] += self.siz[y]

        return True
        
    def size(self, x):
        return self.siz[self.root(x)]

def main():
    N, M = map(int, input().split())
    edge = []
    for _ in range(M):
        s, d, c = map(int, input().split())
        edge.append([s-1, d-1, c])
    
    edge.sort(key=lambda x: x[2])
    sst = []
    min_cost = 0
    uf = UnionFind(N)
    # make smallest spanning tree
    for i, e in enumerate(edge): 
        s, d, c = e
        if uf.issame(s, d):
            continue
        uf.unite(s, d)
        min_cost += c
        sst.append(i)
    
    output = [0, 0]
    # make smallest spanning tree from G excluding one edge
    for id in sst:
        uf = UnionFind(N)
        counter = 1
        cost = 0
        for i, e in enumerate(edge):
            s, d, c = e
            if id == i:
                continue
            if uf.issame(s, d):
                continue
            uf.unite(s, d)
            cost += c
            counter += 1
        if counter < N or cost > min_cost:
            output[0] += 1
            output[1] += edge[id][2]

    print(*output)
          
if __name__=='__main__':
    main()
